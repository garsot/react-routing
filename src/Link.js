import React, { forwardRef } from 'react'
import PropTypes from 'prop-types'

import HistoryManager from './HistoryManager'

/**
 * Native anchor wrapper. Instead `href` use property `to`.
 */
const Link = forwardRef(({ to, children, onClick, ...props }, ref) => {
    return <a ref={ref} href={to} onClick={(e) => { e.preventDefault(); HistoryManager.push(to); onClick && onClick(e) }} {...props}>{children}</a>
})

Link.displayName = 'Link'

Link.propTypes = {
    to: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.shape({
            pathname: PropTypes.string,
            search: PropTypes.object,
            hash: PropTypes.string
        })
    ]).isRequired    
}

export default Link
