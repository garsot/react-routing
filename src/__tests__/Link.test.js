import React from 'react'
import { shallow } from 'enzyme'

import Link from '../Link'

describe('Link tests', () => {      

    beforeEach(() => {
        window.history.pushState({}, null, '/')
    })

    test('should render correctly', () => {

        const output = shallow(<Link to='/test'>Test</Link>)
        expect(output).toMatchSnapshot()
    })

    test('should call the passed onClick handler', () => {

        const handleClick = jest.fn()  

        const output = shallow(<Link to='/test' onClick={handleClick}>Test</Link>)
        output.find('a').simulate('click', { preventDefault: () => { } })
        expect(handleClick).toHaveBeenCalled()
    })

    test('should change history', () => {
        expect(window.location.pathname).toBe('/')

        const output = shallow(<Link to='/test'>Test</Link>)
        output.find('a').simulate('click', { preventDefault: () => { } })

        expect(window.location.pathname).toBe('/test')
    })

})
