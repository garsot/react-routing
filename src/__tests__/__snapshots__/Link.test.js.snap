// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link tests should render correctly 1`] = `
<a
  href="/test"
  onClick={[Function]}
>
  Test
</a>
`;
